File: app/api/restaurants.py
from bson.objectid import ObjectId
from flask import Blueprint, request, current_app as app, session

from app.decorators import api_view, login_required, admin_required

blueprint = Blueprint("api_restaurants", __name__)

@blueprint.route("/", methods=["POST"])
@api_view
@admin_required
def add_restaurant():
    db = app.mongo_client["cs411"]
    document = {
        "name": request.form["name"],
        "rating": float(request.form["rating"]),
        "location": {"type": "Point", "coordinates": [float(request.form["latitude"]), float(request.form["longitude"])]},
        "address": request.form["address"],
        "categories": list(map(str.strip, request.form["categories"].split(","))),
        "yelp_url": request.form["yelp_url"]
    }
    db.restaurants.insert_one(document)
    return {"message": "Success!"}

@blueprint.route("/<id>", methods=["POST"])
@api_view
@admin_required
def update_restaurant(id):
    db = app.mongo_client["cs411"]
    query = {
        "$set": {
            "name": request.form["name"],
            "rating": float(request.form["rating"]),
            "location": {"type": "Point", "coordinates": [float(request.form["latitude"]), float(request.form["longitude"])]},
            "address": request.form["address"],
            "categories": list(map(str.strip, request.form["categories"].split(","))),
            "yelp_url": request.form["yelp_url"]
        }
    }
    db.restaurants.update_one({"_id": ObjectId(id)}, query)
    return {"message": "Success!"}


@blueprint.route("/<id>", methods=["DELETE"])
@api_view
@admin_required
def delete_restaurant(id):
    db = app.mongo_client["cs411"]
    db.restaurants.delete_one({"_id": ObjectId(id)})
    db.user_restaurants.delete_many({"restaurant_id": ObjectId(id)})
    return {"message": "Success!"}

@blueprint.route("/<id>/favorite", methods=["POST"])
@api_view
@login_required
def favorite_restaurant(id):
    db = app.mongo_client["cs411"]
    is_favorited = db.user_restaurants.find_one({"user_id": session["uid"], "restaurant_id": ObjectId(id)})
    if is_favorited is None:
        db.user_restaurants.insert({
            "user_id": session["uid"],
            "restaurant_id": ObjectId(id)
        })
    else:
        db.user_restaurants.delete_one({
            "user_id": session["uid"],
            "restaurant_id": ObjectId(id)
        })
    return {"status": 200}

File: app/static/js/admin/restaurants.js
$("form[action='update']").submit(function(e) {
    e.preventDefault();

    let form = $(this);
    let restaurant_id = form.attr("id");
    let name = $(`#${restaurant_id}-name`).val();
    let address = $(`#${restaurant_id}-addr`).val();
    let rating = parseFloat($(`#${restaurant_id}-rating`).val());
    let latitude = parseFloat($(`#${restaurant_id}-lat`).val());
    let longitude = parseFloat($(`#${restaurant_id}-long`).val());
    let yelp_url = $(`#${restaurant_id}-url`).val();
    let categories = $(`#${restaurant_id}-categories`).val();
    $(`#${restaurant_id}-submit`).attr("disabled", "disabled");
    apiCall("POST", `/api/restaurants/${restaurant_id}`, {
        name: name,
        address: address,
        rating: rating,
        latitude: latitude,
        longitude: longitude,
        yelp_url: yelp_url,
        categories: categories
    },
        function(response) {
            $(`#${restaurant_id}-submit`).removeAttr("disabled", "disabled");
            window.location = "/admin/restaurants";
        },
        function(response) {
            $(`#${restaurant_id}-submit`).removeAttr("disabled", "disabled");
        }
    );
});

$("#add-form").submit(function(e) {
    e.preventDefault();

    let name = $(`#add-name`).val();
    let address = $(`#add-addr`).val();
    let rating = parseFloat($(`#add-rating`).val());
    let latitude = parseFloat($(`#add-lat`).val());
    let longitude = parseFloat($(`#add-long`).val());
    let yelp_url = $(`#add-url`).val();
    let categories = $(`#add-categories`).val();
    $(`#add-submit`).attr("disabled", "disabled");
    apiCall("POST", `/api/restaurants`, {
        name: name,
        address: address,
        rating: rating,
        latitude: latitude,
        longitude: longitude,
        yelp_url: yelp_url,
        categories: categories
    },
        function(response) {
            $(`#add-submit`).removeAttr("disabled", "disabled");
            window.location = "/admin/restaurants";
        },
        function(response) {
            $(`#add-submit`).removeAttr("disabled", "disabled");
        }
    );
});

function deleteRestaurant(id) {
    console.log(id)
    $(`#${id}-delete`).attr("disabled", "disabled");
    apiCall("DELETE", `/api/restaurants/${id}`, {}, function(response) {
        $(`#${id}-delete`).removeAttr("disabled", "disabled");
        window.location = "/admin/restaurants";
    }, function(response) {
        $(`#${id}-delete`).removeAttr("disabled", "disabled");
    });
}
